framework/src/media/HttpInputDataSource.cpp

```diff
@@ -0,0 +1,250 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************/
+
+#include <tinyara/config.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <debug.h>
+#include <unistd.h>
+#include <assert.h>
+#include <media/HttpInputDataSource.h>
+#include <chrono>
+
+#include "utils/MediaUtils.h"
+#include "HttpStream.h"
+#include "StreamBuffer.h"
+#include "StreamBufferReader.h"
+#include "StreamBufferWriter.h"
+
+namespace media {
+namespace stream {
+
+// Content-Type for audio stream
+static const std::string TAG_CONTENT_TYPE = "Content-Type:";
+static const std::string AAC_MIME_TYPE = "audio/aac";
+static const std::string AACP_MIME_TYPE = "audio/aacp";
+static const std::string MPEG_MIME_TYPE = "audio/mpeg";
+static const std::string OPUS_MIME_TYPE = "audio/opus";
+
+static const std::chrono::seconds WAIT_HEADER_TIMEOUT = std::chrono::seconds(3);
+static const std::chrono::seconds WAIT_DATA_TIMEOUT = std::chrono::seconds(3);
+
+HttpInputDataSource::HttpInputDataSource(const std::string &url)
+	: InputDataSource(), mUrl(url)
+{
+	medvdbg("url: %s\n", mUrl.c_str());
+}
+
+HttpInputDataSource::HttpInputDataSource(const HttpInputDataSource &source)
+	: InputDataSource(source), mUrl(source.mUrl)
+{
+}
+
+HttpInputDataSource &HttpInputDataSource::operator=(const HttpInputDataSource &source)
+{
+	InputDataSource::operator=(source);
+	return *this;
+}
+
+bool HttpInputDataSource::open()
+{
+	medvdbg("HttpInputDataSource::open!\n");
+	// create stream buffer to cache downloading data
+	mStreamBuffer = StreamBuffer::Builder()
+							.setBufferSize(4096)
+							.setThreshold(2048)
+							.build();
+	if (!mStreamBuffer) {
+		meddbg("mStreamBuffer is nullptr!\n");
+		return false;
+	}
+
+	mBufferReader = std::make_shared<StreamBufferReader>(mStreamBuffer);
+	mBufferWriter = std::make_shared<StreamBufferWriter>(mStreamBuffer);
+
+	mHttpStream = HttpStream::create();
+	if (!mHttpStream) {
+		meddbg("mHttpStream is nullptr!\n");
+		return false;
+	}
+
+	std::unique_lock<std::mutex> lock(mMutex);
+	mIsHeaderReceived = false;
+	mIsDataReceived = false;
+
+	mThread = std::thread([this]() {
+		medvdbg("download thread enter!\n");
+		//mHttpStream->addHeader("Icy-MetaData:1"); // not support now
+		mHttpStream->setHeaderCallback(HeaderCallback, (void *)this);
+		mHttpStream->setWriteCallback(WriteCallback, (void *)this);
+		if (!mHttpStream->download(mUrl)) {
+			medwdbg("download failed or terminated!\n");
```

Comment:
in this case, Network error. 
so we should call onPlaybackError

Reply:
I will handle this and other problems you mentioned above.
Thanks for your review! :)

Reply:
Currently, we can not notify error from datasource, let me think about it and update in future.
I left a TODO here.

Reply:
std::thread using fixed size of stack, (as I know 1024) is it enough?

Reply:
I changed std::thread to pthread.  Currently the required stack size is about 5KB (default: 8KB).